ownership/sync-repos.py

```diff
@@ -0,0 +1,255 @@
+    # out archived repos and removed any that are pending a rename on
+    # the spreadsheet.
+    for gh_name in names_only_gh.copy():
+        actions.append({
+            'action': 'add_row',
+            'name': gh_name,
+            'why': "Repo present in Github but not spreadsheet"
+        })
+        names_only_gh.remove(gh_name)
+
+    # Make sure nothing is unaccounted for before presenting recommendations.
+    if names_only_gh or names_only_ss:
+        bail("Failed to convert all mismatches between Github and spreadsheet "
+             "into actions! Github remaining = %s; spreadsheet remaining = %s"
+             % (names_only_gh, names_only_ss))
+
+    # Ensure actions don't have any internal conflicts. It might be OK
+    # if the same *name* comes up twice; for example, we might want to
+    # move an archived repo to the unspported org *and* remove it from
+    # the spreadsheet. However, two actions for the same row ID is a
+    # problem.
+    seen_row_ids = set()
+    for action in actions:
+        row_id = action.get('row_id')
+        if not row_id:
+            continue
+        if row_id in seen_row_ids:
+            bail("Possible bug in script: Was going to suggest multiple "
+                 "actions for row %s: %s"
+                 % (row_id, [r for r in actions if r.get('row_id') == row_id]))
+        seen_row_ids.add(row_id)
+
+    # For now, just make recommendations, don't make changes.
+    if actions:
+        print("Recommendations:")
+
+        groupkey = lambda a: (a['action'], a['why'])
+        for ((action_name, why), group) in groupby(sorted(actions, key=groupkey), groupkey):
+            print()
+            print("Action: %s -- %s" % (action_name, why))
+            for action in group:
+                data = action.copy()
+                del data['action']
+                del data['name']
+                del data['why']
+                if data:
+                    data_str = " " + str(data)
+                else:
+                    data_str = ""
+                print("- %s%s" % (action['name'], data_str))
+    else:
+        print("No changes required. ✔")
+
+
+def fetch_github(hub, orgs):
+    """
+    Generator of repos found in Github orgs, as dicts:
+
+    - name: Lowercased full name of repository
+    - archived: Boolean, true if archived
+    """
+    for org in orgs:
+        for repo in hub.organization(org).repositories():
+            yield {'name': str(repo).lower(),
+                   'archived': repo.archived}
+
+def fetch_spreadsheet(spreadsheet_csv, orgs):
+    """
+    Generator for repos found in spreadsheet, as dicts:
+
+    - name: Lowercased full name of repository
+    - row_id: Spreadsheet row number
+
+    Raises an exception if any entries are not in provided orgs.
+    """
+    if spreadsheet_csv:
+        with open(spreadsheet_csv, 'r') as csvf:
+            reader = csv.reader(csvf)
+            # Figure out which columns are which
```

Comment:
I'm sure there's a better way to do this, but we'll be switching to reading from Google's Spreadsheet API instead, so whatever.

Reply:
The dict reader is good for that and then you don't have to care so much about column indices but agree that you don't need to fix it since this is temporary.